catalog/core/catalog-core-validator/src/test/groovy/ddf/catalog/validation/impl/RelationshipValidatorSpec.groovy

```diff
@@ -0,0 +1,127 @@
+        "sourceattribute" | "sourcevalue" | CANNOT_HAVE  | "targetattribute" | "val1"     | "val2"
+        "sourceattribute" | "sourcevalue" | MUST_HAVE    | "mismatch"        | "val1"     | "val2"
+        "sourceattribute" | "sourcevalue" | MUST_HAVE    | "targetattribute" | "val1"     | "mismatch"
+    }
+
+    @Unroll
+    def "validation permutations"(
+            boolean expectViolation,
+            String sourceAttribute,
+            String sourceValue,
+            String relationship,
+            String targetAttribute,
+            String... targetValues) {
+        setAttribute("source.undef", null)
+        setAttribute("source.def", "blah blah")
+        setAttribute("target.undef", null)
+        setAttribute("target.def", "blah blah")
+        setAttribute("source.multivalued", "one", "two", "three")
+        setAttribute("target.multivalued", "one", "two", "three")
+        RelationshipValidator validator =
+                new RelationshipValidator(
+                        sourceAttribute, sourceValue, relationship, targetAttribute, targetValues)
+        when:
+        Optional<MetacardValidationReport> metacardValidationReport =
+                validator.validateMetacard(metacard)
+        then:
+        expectViolation == (metacardValidationReport.isPresent() && !metacardValidationReport.get().getAttributeValidationViolations().isEmpty())
+        when:
+        ValidationException e
+        try {
+            validator.validate(metacard)
+        } catch (ValidationException v) {
+            e = v
+        }
+        then:
+        expectViolation == (e != null)
+        where:
+        expectViolation | sourceAttribute | sourceValue | relationship  | targetAttribute      | targetValues
+        false           | "source.undef"  | null        | MUST_HAVE     | "target.def"         | null
+        true            | "source.def"    | null        | MUST_HAVE     | "target.undef"       | null
+        false           | "source.def"    | null        | MUST_HAVE     | "target.def"         | ["blah blah"]
+        false           | "source.def"    | "blah"      | MUST_HAVE     | "target.undef"       | null
+        true            | "source.def"    | "blah blah" | MUST_HAVE     | "target.undef"       | null
+        true            | "source.def"    | null        | CANNOT_HAVE   | "target.multivalued" | ["two"]
+        true            | "source.def"    | null        | CANNOT_HAVE   | "target.def"         | ["blah blah"]
+        false           | "source.def"    | null        | CANNOT_HAVE   | "target.multivalued" | ["five"]
+        false           | "source.def"    | null        | CAN_ONLY_HAVE | "target.multivalued" | ["one", "two", "three", "four"]
+        true            | "source.def"    | null        | CAN_ONLY_HAVE | "target.multivalued" | ["one", "two", "five"]
+    }
+
+    void setAttribute(String attribute, String... value) {
+        metacard = metacard == null ? new MetacardImpl() : metacard
```

Comment:
✏️ Could instantiate `metacard` where it is defined instead.

Reply:
Refactoring cruft. 👍